### PR TITLE
Implement `cudf-polars` chunked parquet reading

### LINKED ISSUE #16818: [FEA] Investigate the chunked parquet reader for Polars GPU engine
Some users experience out-of-memory errors during IO when loading datasets that they _feel_ like should be fine for their given GPU. This is currently less significant for cudf.pandas, as we now enable a prefetch-optimized unified memory by default.

Because we don't currently have a similar UVM setup for the Polars GPU engine, this is an acute pain point that blocks usage for many workflows. We've developed chunked readers for Parquet and ORC files that may be able to help in this situation.

Initial testing suggests that a properly configured chunked parquet reader may be effective at reducing peak memory requirements without significantly impacting performance.

For example, running PDS-H q7 at SF200 immediately runs into an OOM with the default Parquet reader. With a `pass_read_limit` of 16GB for the chunked reader, we can smoothly finish the query and provide a speedup with an H100 vs. the CPU engine on a high-end CPU.

Default CPU engine on a dual socket Intel 8480CL:
```
Running experiments...
Running scale factor 200.0 on GPU setting false...
Code block 'Run polars query 7' took: 18.65865 s
Experiments completed. ...
Run complete!
```

Default GPU engine behavior with cuda-async memory resource
```
Running experiments...
Running scale factor 200.0 on GPU setting true...
Code block 'Run polars-gpu-cuda-async query 7' took: 3.28927 s
q7 FAILED
MemoryError: std::bad_alloc: out_of_memory: CUDA error at: /tmp/pip-build-env-7ci1mf7i/normal/lib/python3.11/site-packages/librmm/include/rmm/mr/device/cuda_async_view_memory_resource.hpp:111: cudaErrorMemoryAllocation out of memory
Experiments completed. ...
Run complete!
```

GPU engine behavior with cuda-async memory resource and `"pass_read_limit": 16024000000`
```
Running experiments...
Running scale factor 200.0 on GPU setting true...
Code block 'Run polars-gpu-cuda-async query 7' took: 10.78470 s
Experiments completed. ...
Run complete!
```

We should do a full evaluation of the chunked Parquet reader on the [PDS-H benchmarks](https://github.com/pola-rs/polars-benchmark) to empirically assess the potential opportunity and tradeoffs for chunked IO. Starting with Parquet makes sense, as it's a more common file format in the PyData world. We can expand from there, as needed.

cc @quasiben @brandon-b-miller (offline discussion)

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR provides access to the libcudf chunked parquet reader through the `cudf-polars` gpu engine, inspired by the cuDF python implementation. 

Closes #16818

### CODE REVIEW

File: python/cudf_polars/cudf_polars/callback.py [hunk 1]
```diff
@@ -174,7 +175,9 @@ def execute_with_cudf(
     device = config.device
     memory_resource = config.memory_resource
     raise_on_fail = config.config.get("raise_on_fail", False)
-    if unsupported := (config.config.keys() - {"raise_on_fail"}):
+    parquet_options = config.config.get("parquet_options", {})
+    cudf_polars.dsl.translate.ir.parquet_options = parquet_options
```
Comment (brandon-b-miller) [inline python/cudf_polars/cudf_polars/callback.py:179]: This was just a quick hack to get things working, but we should decide on a way we're happy with for configuring the reader as well as possibly other configurable bits. 

High level questions:
- Should we have a global config, or a per query config (per GPU-Engine) 
- If per query, how should we pass it around? Should `translate_ir` take another argument and make it available to each overload of `_translate_ir`? 


cc @wence-

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 2]
```diff
@@ -339,17 +340,46 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
...
+                concatenated_columns = tbl.columns()
+                while reader.has_next():
+                    tbl = reader.read_chunk().tbl
+
+                    for i in range(tbl.num_columns()):
+                        concatenated_columns[i] = plc.concatenate.concatenate(
+                            [concatenated_columns[i], tbl._columns[i]]
+                        )
+                        # Drop residual columns to save memory
+                        tbl._columns[i] = None
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/ir.py:364]: This would do fewer copies if we did something like:
```
tables = []
chunk = reader.read_chunk()
tables.append(chunk.tbl)
names = chunk.column_names()
while reader.has_next():
   tables.append(reader.read_chunk().tbl)
# deliberately overwriting so we free the chunk tables
tables = plc.concatenate.concatenate(tables)
df = DataFrame.from_table(tables, names=names)
```

Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:364]: It would be really nice if `reader` had a `column_names` method, because then we could do:
```
def tables(reader):
    while reader.has_next():
        yield reader.read_chunk().tbl
names = reader.column_names()
df = DataFrame.from_table(plc.concatenate(list(tables(reader))), names=names)
```
and then we wouldn't need to worry about manually releasing objects.

File: python/cudf_polars/cudf_polars/dsl/translate.py [hunk 3]
```diff
@@ -393,6 +404,9 @@ def translate_ir(visitor: NodeTraverser, *, n: int | None = None) -> ir.IR:
             f"No support for polars IR {version=}"
         )  # pragma: no cover; no such version for now.
 
+    IR._config = config
+    Expr._config = config
+
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/translate.py:409]: Please no! Can we figure out a good way to pass the configuration into `evaluate`, which is where it is needed, rather than setting class attributes.

**Review by wence- — CHANGES_REQUESTED** [review]
Minor tweaks, overall, I think the implementation side looks in good shape.

There's an open question about whether we want to read all the chunks and do a concatenate at the end, but I will wait for benchmarking on that.

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 4]
```diff
@@ -418,17 +428,47 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
...
+                chunks.append(tbl)
+                names = chk.column_names()
+                while reader.has_next():
+                    chunks.append(reader.read_chunk().tbl)
+
+                chunks = plc.concatenate.concatenate(chunks)
+                df = DataFrame.from_table(
+                    chunks,
+                    names=names,
+                )
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/ir.py:458]:
```suggestion
                df = DataFrame.from_table(chunks, names=names)
```

**Comment by wence-** [comment]
We need to put some documentation about the chunked reading somewhere. As a minimum, can you add a new `md` file in `cudf_polars/docs` and talk through what settings are available, and what they mean?

**Comment by brandon-b-miller** [comment]
> We need to put some documentation about the chunked reading somewhere. As a minimum, can you add a new `md` file in `cudf_polars/docs` and talk through what settings are available, and what they mean?

Just realizing now, this is unaddressed. I'll add this.

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 2 above)
Comment (vyasr) [reply python/cudf_polars/cudf_polars/dsl/ir.py:538]: @brandon-b-miller can you summarize the results of all the concatenation experiments here so that we have a record of what decision was made and why?

Comment (brandon-b-miller) [reply python/cudf_polars/cudf_polars/dsl/ir.py:538]: Yes - basically here we tried varying the way we concat things together in the end from less concatting of larger chunks to more concatting of smaller ones and now lean towards the second approach. Larger concats appear to provide a marginal performance improvement while also introducing significant instability at larger dataset sizes which didn't seem to be worth it. Something like a few % higher throughput for large datasets at the expense of OOM-ing a bunch more elsewhere.

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 5]
```diff
@@ -498,25 +509,59 @@ def do_evaluate(
...
-            )
-            df = DataFrame.from_table(
-                tbl_w_meta.tbl,
-                # TODO: consider nested column names?
-                tbl_w_meta.column_names(include_children=False),
-            )
-            if filters is not None:
-                # Mask must have been applied.
-                return df
+            parquet_options = config.config.get("parquet_options", {})
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/ir.py:512]: Can we validate the provided options? We've already checked that parquet_options was valid. But can we also ensure `{"chunked", "chunk_read_limit", "pass_read_limit"}.issuperset(parquet_options)`?

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# GPUEngine Configuration Options
+
+The `polars.GPUEngine` object may be configured in several different ways.
+
+## Parquet Reader Options
+
+Chunked reading is controlled by passing a dictionary of options to the `GPUEngine` object. Details may be found following the links to the underlying `libcudf` reader.
+- `parquet_chunked`, indicicates is chunked parquet reading is to be used, default True.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:8]:
```suggestion
- `chunked`, indicates if chunked parquet reading is to be used, default False.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+- [chunk_read_limit](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum size per chunk, default unlimited.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:9]:
```suggestion
- [`chunk_read_limit`](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum size per chunk, default unlimited.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+- [pass_read_limit](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum memory used for decompression, default 16GiB.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:10]:
```suggestion
- [`pass_read_limit`](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum memory used for decompression, default 16GiB.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+For example, one would pass these parameters as follows:
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:12]:
```suggestion
For example, to select the chunked reader with custom values for `pass_read_limit` and `chunk_read_limit`:
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+```python
+engine = GPUEngine(
+    raise_on_fail=True,
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:15]:
```suggestion
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    parquet_options={
+        'parquet_chunked': True,
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:17]:
```suggestion
        'chunked': True,
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        'chunk_read_limit': int(1e9),
+        'pass_read_limit': int(4e9)
+    }
+)
+result = query.collect(engine=engine)
+```
+Note that passing `parquet_chunked: False` disables chunked reading entirely, and thus `chunk_read_limit` and `pass_read_limit` will have no effect.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:24]:
```suggestion
Note that passing `chunked: False` disables chunked reading entirely, and thus `chunk_read_limit` and `pass_read_limit` will have no effect.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# GPUEngine Configuration Options
+
+The `polars.GPUEngine` object may be configured in several different ways.
+
+## Parquet Reader Options
+
+Chunked reading is controlled by passing a dictionary of options to the `GPUEngine` object. Details may be found following the links to the underlying `libcudf` reader.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:7]: *nit*: This doesn't tell me why I should use "chunked" reading.

How about:

> Reading large parquet files, especially if they are compressed, can run out of device memory. At the cost of a small drop in performance, the memory used for parquet reads can be reduced by selecting the "chunked" reader. When enabled, parquet files are read in chunks with a limit on the amount of memory that is used.

> To configure the parquet reader, we provide a dictionary of options to the `parquet_options` keyword in the `GPUEngine` object. Valid keys and values are:

...

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 14]
```diff
@@ -294,6 +301,9 @@ class Scan(IR):
     predicate: expr.NamedExpr | None
     """Mask to apply to the read dataframe."""
 
+    PARQUET_DEFAULT_CHUNK_SIZE: int = 0
+    PARQUET_DEFAULT_PASS_LIMIT: int = 17179869184  # 16GiB
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/ir.py:305]:
```suggestion
    PARQUET_DEFAULT_PASS_LIMIT: int = 16 * 1024**3  # 16GiB
```

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 15]
```diff
@@ -294,6 +301,9 @@ class Scan(IR):
     predicate: expr.NamedExpr | None
     """Mask to apply to the read dataframe."""
 
+    PARQUET_DEFAULT_CHUNK_SIZE: int = 0
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/ir.py:304]:
```suggestion
    PARQUET_DEFAULT_CHUNK_SIZE: int = 0  # unlimited
```

File: python/cudf_polars/cudf_polars/testing/asserts.py [hunk 16]
```diff
@@ -19,10 +19,13 @@
 
 __all__: list[str] = ["assert_gpu_result_equal", "assert_ir_translation_raises"]
 
+DEFAULT_GPU_ENGINE: GPUEngine = GPUEngine(raise_on_fail=True)
+
 
 def assert_gpu_result_equal(
     lazydf: pl.LazyFrame,
     *,
+    engine: GPUEngine = DEFAULT_GPU_ENGINE,
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/testing/asserts.py:28]: Can we shovel this in to the function? That is, `engine: GPUEngine | None = None` and `engine = GPUEngine(raise_on_fail=True) if engine is None else engine`
?

**Review by wence- — COMMENTED** [review]
Thanks Brandon, looks good. I had some doc suggestions, because it looks like the docs don't agree with the code in terms of what the defaults are.

Can you clarify what we are picking as the default mode?

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
+# GPUEngine Configuration Options
+
+The `polars.GPUEngine` object may be configured in several different ways.
+
+## Parquet Reader Options
+Reading large parquet files incurs significant memory overhead, especially when the files are compressed. This may lead to out of memory errors for some workflows. To mitigate this, the "chunked" parquet reader may be selected. When enabled, parquet files are read in chunks with a limit on the amount of memory that is used, at the cost of a small drop in performance.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:6]:
```suggestion
Reading large parquet files can use a large amount of memory, especially when the files are compressed. This may lead to out of memory errors for some workflows. To mitigate this, the "chunked" parquet reader may be selected. When enabled, parquet files are read in chunks with a limit on the amount of memory that is used, at the cost of a small drop in performance.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+
+To configure the parquet reader, we provide a dictionary of options to the `parquet_options` keyword of the `GPUEngine` object. Valid keys and values are:
+- `chunked`, indicicates is chunked parquet reading is to be used, default True.
```
Comment (wence-) [inline docs/cudf/source/cudf_polars/engine_options.md:10]:
```suggestion
- `chunked`, indicates is chunked parquet reading is to be used, default True.
```

Comment (wence-) [reply docs/cudf/source/cudf_polars/engine_options.md:10]: In the code, the default is still False. Which is it to be?

Comment (brandon-b-miller) [reply docs/cudf/source/cudf_polars/engine_options.md]: True.

File: python/cudf_polars/tests/test_scan.py [hunk 19]
```diff
@@ -13,6 +13,8 @@
     assert_ir_translation_raises,
 )
 
+NO_CHUNK_ENGINE = pl.GPUEngine(raise_on_fail=True, parquet_options={"chunked": False})
+
```
Comment (wence-) [inline python/cudf_polars/tests/test_scan.py:17]: *nit*: Should we just test the chunked reader and the non-chunked reader for all parquet tests? I would prefer yes, since the route to applying filters is different (for example).

I.e. parametrize the format with `parquet` and `chunked-parquet` and use that to determined the `"chunked"` parameter.

Comment (brandon-b-miller) [reply python/cudf_polars/tests/test_scan.py:17]: yeah, no problem.

Comment (brandon-b-miller) [reply python/cudf_polars/tests/test_scan.py:17]: Concretely, we'd error if `skip_rows` is passed to `pl.scan_parquet`. But allow `.collect(engine=GPUEngine(parquet_options={'chunked': False, 'skip_rows': 42})`

Comment (wence-) [reply python/cudf_polars/tests/test_scan.py:17]: No, `skip_rows` comes from the polars IR, we can't ask the user to pass it in like that.

`skip_rows` is exposed for the chunked reader in pylibcudf and seems to be implemented in the C++. Can you report a bug that it doesn't work?

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 20]
```diff
@@ -498,25 +509,59 @@ def do_evaluate(
...
-            if filters is not None:
-                # Mask must have been applied.
-                return df
+            parquet_options = config.config.get("parquet_options", {})
+            if parquet_options.get("chunked", True):
+                reader = plc.io.parquet.ChunkedParquetReader(
+                    plc.io.SourceInfo(paths),
+                    columns=with_columns,
+                    nrows=n_rows,
+                    skip_rows=skip_rows,
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/ir.py:518]: Concretely, why does `skip_rows != 0` not work? If there is a simple repro, it is a bug.

Comment (brandon-b-miller) [reply python/cudf_polars/cudf_polars/dsl/ir.py:518]:
```python
import pylibcudf as plc
import pyarrow as pa
import pyarrow.parquet as pq

path = "./test.parquet"
pq.write_table(pa.Table.from_pydict({'a':[1,2,3]}), path)

reader = plc.io.parquet.ChunkedParquetReader(
        plc.io.SourceInfo([path]),
        columns=['a'],
        nrows=-1,
        skip_rows=2,
        chunk_read_limit=0,
        pass_read_limit=17179869184 # 16 GiB
)

```

```
RuntimeError: CUDF failure at: /xyz/cudf/cpp/src/io/parquet/reader.cpp:44: skip_rows > 0 is not currently supported in the Chunked Parquet reader.
```

Comment (brandon-b-miller) [reply python/cudf_polars/cudf_polars/dsl/ir.py:518]: @wence- I believe this issue is coming up through the tests that slice a scanned parquet file. The default `n_rows` value of `None` should lead to a `skip_rows` value of zero in most cases, so I think we should be good there. 

I think our impl is pushing the slice down into a skip_rows operation. If we could just not permit that unless the chunked reader is disabled, we might be ok, what do you think?

**Comment by brandon-b-miller** [comment]
In addition to the slicing issue, switching to chunked reading by default seems to shake out another chunked parquet reader bug with `n_rows > 0` 

<details>

```python
import pylibcudf as plc
import pyarrow as pa
import pyarrow.parquet as pq


data = {
            "a": [1, 2, 3, None, 4, 5],
            "b": ["ẅ", "x", "y", "z", "123", "abcd"],
            "c": [None, None, 4, 5, -1, 0],
        }


path = "./test.parquet"
pq.write_table(pa.Table.from_pydict(data), path)

reader = plc.io.parquet.ChunkedParquetReader(
        plc.io.SourceInfo([path]),
        columns=['a', 'b', 'c'],
        nrows=2,
        skip_rows=0,
        chunk_read_limit=0,
        pass_read_limit=17179869184 # 16 GiB
)
chk = reader.read_chunk()
tbl = chk.tbl
names = chk.column_names()
concatenated_columns = tbl.columns()
while reader.has_next():
    tbl = reader.read_chunk().tbl

    for i in range(tbl.num_columns()):
        concatenated_columns[i] = plc.concatenate.concatenate(
            [concatenated_columns[i], tbl._columns[i]]
        )
        # Drop residual columns to save memory
        tbl._columns[i] = None

gpu_result = plc.interop.to_arrow(tbl)
cpu_result = pq.read_table(path)[:2]

print(cpu_result.column(1).to_pylist())
print(gpu_result.column(1).to_pylist())
```

</details>

this yields 

```
['ẅ', 'x']
['ẅ', 'x\x00\x00\x00\x00\x00\x00\x00\x00\x00']
```

I suppose this is a separate issue from https://github.com/rapidsai/cudf/issues/17158 since the trunk includes this fix now I believe.

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 20 above)
Comment (vyasr) [reply python/cudf_polars/cudf_polars/dsl/ir.py:518]: See #16186

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 21]
```diff
@@ -349,10 +349,10 @@ def __init__(
             # TODO: polars has this implemented for parquet,
             # maybe we can do this too?
             raise NotImplementedError("slice pushdown for negative slices")
-        if self.typ == "csv" and self.skip_rows != 0:  # pragma: no cover
+        if self.typ in {"csv", "parquet"} and self.skip_rows != 0:  # pragma: no cover
             # This comes from slice pushdown, but that
             # optimization doesn't happen right now
-            raise NotImplementedError("skipping rows in CSV reader")
+            raise NotImplementedError("skipping rows in CSV or Parquet reader")
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py]: OK, I pushed an implementation of option 3.

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 20 above)
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py]: OK, added a workaround.

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 22]
```diff
@@ -0,0 +1,25 @@
+# GPUEngine Configuration Options
+
+The `polars.GPUEngine` object may be configured in several different ways.
+
+## Parquet Reader Options
+Reading large parquet files can use a large amount of memory, especially when the files are compressed. This may lead to out of memory errors for some workflows. To mitigate this, the "chunked" parquet reader may be selected. When enabled, parquet files are read in chunks with a limit on the amount of memory that is used, at the cost of a small drop in performance.
```
Comment (vyasr) [inline docs/cudf/source/cudf_polars/engine_options.md:6]:
```suggestion
Reading large parquet files can use a large amount of memory, especially when the files are compressed. This may lead to out of memory errors for some workflows. To mitigate this, the "chunked" parquet reader may be selected. When enabled, parquet files are read in chunks, limiting the peak memory usage at the cost of a small drop in performance.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+
+
+To configure the parquet reader, we provide a dictionary of options to the `parquet_options` keyword of the `GPUEngine` object. Valid keys and values are:
+- `chunked`, indicates is chunked parquet reading is to be used, default True.
```
Comment (vyasr) [inline docs/cudf/source/cudf_polars/engine_options.md:10]: Let's write the defaults here in sentences rather than like docstrings.
```suggestion
- `chunked` indicates that chunked parquet reading is to be used. By default, chunked reading is turned on.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+- [`chunk_read_limit`](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum size per chunk, default unlimited.
```
Comment (vyasr) [inline docs/cudf/source/cudf_polars/engine_options.md:11]:
```suggestion
- [`chunk_read_limit`](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum size per chunk. By default, the maximum chunk size is unlimited.
```

File: docs/cudf/source/cudf_polars/engine_options.md [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+- [`pass_read_limit`](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum memory used for decompression, default 16GiB.
```
Comment (vyasr) [inline docs/cudf/source/cudf_polars/engine_options.md:12]:
```suggestion
- [`pass_read_limit`](https://docs.rapids.ai/api/libcudf/legacy/classcudf_1_1io_1_1chunked__parquet__reader#aad118178b7536b7966e3325ae1143a1a) controls the maximum memory used for decompression. The default pass read limit is 16GiB.
```

File: python/cudf_polars/cudf_polars/callback.py [hunk 26]
```diff
@@ -145,7 +146,31 @@ def _callback(
...
+        Configuration options to validate.
+
+    Raises
+    ------
+    ValueError
+        If the configuration contains unsupported options.
+    """
+    if unsupported := (config.keys() - {"raise_on_fail", "parquet_options"}):
+        raise ValueError(
+            f"Engine configuration contains unsupported settings {unsupported}"
```
Comment (vyasr) [inline python/cudf_polars/cudf_polars/callback.py:168]:
```suggestion
            f"Engine configuration contains unsupported settings: {unsupported}"
```

File: python/cudf_polars/cudf_polars/callback.py [hunk 27]
```diff
@@ -145,7 +146,31 @@ def _callback(
...
+        )
+    if parquet_options := config.get("parquet_options", {}):
+        assert {"chunked", "chunk_read_limit", "pass_read_limit"}.issuperset(
+            parquet_options
+        )
```
Comment (vyasr) [inline python/cudf_polars/cudf_polars/callback.py:173]: The conditional isn't needed here since you default to the empty set anyway right?
```suggestion
    assert {"chunked", "chunk_read_limit", "pass_read_limit"}.issuperset(
        config.get("parquet_options", {})
    )
```

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 28]
```diff
@@ -498,25 +509,77 @@ def do_evaluate(
...
+                )
+                chk = reader.read_chunk()
+                rows_left_to_skip = skip_rows
+
+                def slice_skip(tbl: plc.Table):
+                    nonlocal rows_left_to_skip
+                    if rows_left_to_skip > 0:
+                        table_rows = tbl.num_rows()
+                        chunk_skip = min(rows_left_to_skip, table_rows)
+                        (tbl,) = plc.copying.slice(tbl, [chunk_skip, table_rows])
```
Comment (vyasr) [inline python/cudf_polars/cudf_polars/dsl/ir.py:540]: I'm not sure if there's meaningful overhead in calling slice when `chunk_skip == table_rows`. If there is we might want to elide the call. That's a minor optimization though and I don't think that it's worth experimenting with now. Maybe just adding a comment for future experimentation.
```suggestion
                        # TODO: Check performance impact of skipping this
                        # call and creating an empty table manually when the
                        # slice would be empty (chunk_skip == table_rows).
                        (tbl,) = plc.copying.slice(tbl, [chunk_skip, table_rows])
```

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 29]
```diff
@@ -498,25 +509,77 @@ def do_evaluate(
...
+                    tbl = slice_skip(reader.read_chunk().tbl)
+
+                    for i in range(tbl.num_columns()):
+                        concatenated_columns[i] = plc.concatenate.concatenate(
+                            [concatenated_columns[i], tbl._columns[i]]
+                        )
+                        # Drop residual columns to save memory
+                        tbl._columns[i] = None
+
+                df = DataFrame.from_table(
```
Comment (vyasr) [inline python/cudf_polars/cudf_polars/dsl/ir.py:558]: Does it make sense to return early here since we don't support predicates anyway right now? Or does it not really matter since everything after is a no-op anyway.

File: python/cudf_polars/tests/test_scan.py (hunk 19 above)
Comment (vyasr) [reply python/cudf_polars/tests/test_scan.py:17]: Lawrence put in a fix in https://github.com/rapidsai/cudf/pull/16944/commits/df341ea408c76d29a2456493be9644fbd71abecd.

**Review by vyasr — APPROVED** [review]
I have some suggestions to improve the docs and a couple of minor questions on implementation, but nothing blocking. I think that this is good to merge when you are happy.

File: python/cudf_polars/cudf_polars/dsl/ir.py (hunk 29 above)
Comment (brandon-b-miller) [reply python/cudf_polars/cudf_polars/dsl/ir.py:561]: Certain operations like filtering with an expression still invoke the predicate codepath at the very end, so we need the data to reach that point to get the right answer. `test_parquet_filters::test_scan_by_hand` is a good example.

Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:561]: Yeah. Post 24.12 I am going to refactor the ir nodes (and especially IO) to be easier to follow

**Review by wence- — APPROVED** [review]

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 30]
```diff
@@ -208,8 +214,9 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
             translation phase should fail earlier.
         """
         return self.do_evaluate(
+            config,
```
Comment (rjzamora) [inline python/cudf_polars/cudf_polars/dsl/ir.py:217]: @wence- - Just a note. Pretty sure this means we will need to pass in a config object to every single task in the task graph for multi-gpu.

Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:217]: Argh, ok, painful. Let's try and figure something out (especially because the config object can contain a memory resource).

Comment (rjzamora) [reply python/cudf_polars/cudf_polars/dsl/ir.py:217]: Would it make sense for `config` to be a required IR constructor argument, and **not** require it as an argument to `do_evaluate` (unless necessary)?

Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/ir.py:217]: Ah, we could pass the config options we need into the `Scan` node during `translate`, and then it never needs to be in `do_evaluate` at all

Comment (rjzamora) [reply python/cudf_polars/cudf_polars/dsl/ir.py:217]: Possible revision proposed here: https://github.com/rapidsai/cudf/pull/17339
